### PR TITLE
Client side throttling

### PR DESCRIPTION
**Pull request by kriscode1**
For https://github.com/coreweave/cloud-app/issues/2207

I'll open a cloud-app PR to update the dependency in the dockerfile to resolve the issue.

### CODE REVIEW

File: kopf/_cogs/configs/configuration.py [hunk 1]
```diff
@@ -352,6 +352,14 @@ class NetworkingSettings:
     For more information on the API errors retrying, see :doc:`api-retrying`.
     """
 
+    burst_concurrency: Optional[int] = None
+    """
+    Maximum number of requests that may be made at a given time.
```
Comment (mecampbellsoup) [inline kopf/_cogs/configs/configuration.py:357]: "Maximum number of unresolved requests that may exist concurrently"?

File: kopf/_core/reactor/running.py [hunk 2]
```diff
@@ -515,6 +515,11 @@ async def _startup_cleanup_activities(
         logger.warning("Startup activity is only partially executed due to cancellation.")
         raise
 
+    if settings.networking.burst_concurrency is not None:
+        api.burst_limiter = asyncio.BoundedSemaphore(settings.networking.burst_concurrency)
```
Comment (mecampbellsoup) [inline kopf/_core/reactor/running.py:519]: https://docs.python.org/3.11/library/asyncio-sync.html#asyncio.BoundedSemaphore

> class asyncio.BoundedSemaphore(value=1)
> A bounded semaphore object. Not thread-safe.

> Bounded Semaphore is a version of [Semaphore](https://github.com/coreweave/kopf/pull/1/commits/55f3caa0c44d40514bf862db3607f1424c1e2e2f#asyncio.Semaphore) that raises a [ValueError](https://github.com/coreweave/kopf/pull/1/commits/exceptions.html#ValueError) in [release()](https://github.com/coreweave/kopf/pull/1/commits/55f3caa0c44d40514bf862db3607f1424c1e2e2f#asyncio.Semaphore.release) if it increases the internal counter above the initial value.

So does the code need to handle potential `ValueError` raised then, or are you assuming the code will always correctly check the sempahore's counter before attempting to acquire a slot?

Comment (mecampbellsoup) [reply kopf/_core/reactor/running.py:519]: In other words, will [this line](https://github.com/coreweave/kopf/pull/1/commits/55f3caa0c44d40514bf862db3607f1424c1e2e2f#diff-feb5f136ad9f387915f74f048287c7e84d68e8ac0f51b57184efeb901a031e69R45) raise an exception if calling code tries to use the semaphore, or does the context manager handle the "you have to wait for a slot in this semaphore" bit?

**Review by mecampbellsoup — APPROVED** [review]
Very nice, I like it.

File: kopf/_core/reactor/running.py (hunk 2 above)
Comment (kriscode1) [reply kopf/_core/reactor/running.py:519]: > So does the code need to handle potential `ValueError` raised then, or are you assuming the code will always correctly check the sempahore's counter before attempting to acquire a slot?

Yes the code is written such that we should never see a `ValueError`.

> In other words, will [this line](https://github.com/coreweave/kopf/pull/1/commits/55f3caa0c44d40514bf862db3607f1424c1e2e2f#diff-feb5f136ad9f387915f74f048287c7e84d68e8ac0f51b57184efeb901a031e69R45) raise an exception if calling code tries to use the semaphore, or does the context manager handle the "you have to wait for a slot in this semaphore" bit?

The context manager does the work of acquiring and releasing so there will always be one release for every one acquire.

**Pull request merged**